ui/components-react/src/test/tree/controlled/TreeEventDispatcher.test.ts

```diff
@@ -3,604 +3,823 @@
+        }
+      );
 
-      it("replaces selected nodes using range selection from one node", async () => {
-        const selection = {
-          from: deselectedNodes[0].id,
-          to: deselectedNodes[0].id,
-        };
+      dispatcher.onNodeClicked("A", {} as any);
+      expect(treeEvents.onSelectionReplaced).to.be.calledOnce;
 
-        const spy = sinon.spy();
-        treeEventsMock.setup((x) => x.onSelectionReplaced).returns(() => spy);
-        selectionManager.onSelectionReplaced.emit({
-          selectedNodeIds: selection,
-        });
-        expect(spy).to.be.called;
-
-        const spyArgs = spy.args[0][0] as TreeSelectionReplacementEventArgs;
-        const results = await extractSequence(rxjsFrom(spyArgs.replacements));
-        expect(results).to.not.be.empty;
-        const selectionChange = results[0];
-        expect(selectionChange.selectedNodeItems).to.be.deep.eq([
-          deselectedNodes[0].item,
+      dispatcher.onNodeClicked("A", { shiftKey: true } as any);
+      expect(treeEvents.onSelectionReplaced).to.be.calledTwice;
+      const [args1] = treeEvents.onSelectionReplaced.args[1];
+      expect(await extractSequence(toRxjsObservable(args1.replacements)))
+        .to.have.lengthOf(1)
+        .and.containSubset([
+          {
+            selectedNodeItems: [{ id: "A" }],
+          },
         ]);
-      });
     });
   });
 
-  describe("onNodeCheckboxClicked", () => {
-    it("changes state for clicked node", async () => {
-      const expectedAffectedNodeItems = [deselectedNodes[0].item];
+  describe("drag selection", () => {
+    it("selects nodes", async () => {
+      const { dispatcher } = setupTreeEventDispatcher(
+        SelectionMode.Multiple,
+        (model) => {
+          model.setChildren(
+            undefined,
+            [
+              createTreeNodeInput("A"),
+              createTreeNodeInput("B"),
+              createTreeNodeInput("C"),
+            ],
+            0
+          );
+        }
+      );
+
+      dispatcher.onNodeMouseDown("A");
+
+      expect(treeEvents.onSelectionModified).to.be.calledOnce;
+      const [args0] = treeEvents.onSelectionModified.args[0];
+      const result0 = startExtractingSequence(
+        toRxjsObservable(args0.modifications)
+      );
 
-      const spy = sinon.spy();
-      treeEventsMock.setup((x) => x.onCheckboxStateChanged).returns(() => spy);
+      dispatcher.onNodeMouseMove("B");
+      dispatcher.onNodeMouseMove("C");
+      fireEvent.mouseUp(window);
 
-      dispatcher.onNodeCheckboxClicked(deselectedNodes[0].id, CheckBoxState.On);
+      await result0.waitForComplete;
+      expect(result0.current.sequence)
+        .to.have.lengthOf(2)
+        .and.containSubset([
+          {
+            selectedNodeItems: [{ id: "A" }, { id: "B" }],
+            deselectedNodeItems: [],
+          },
+          {
+            selectedNodeItems: [{ id: "C" }],
+            deselectedNodeItems: [],
+          },
```

Comment:
interesting - why C is separate from A and B?

Reply:
This is happening because `SelectionHandler` does not invoke `updateSelection` if there are no selection changes: https://github.com/iTwin/appui/blob/71ab57c9416bb598688d13f47bff3563489d30ac/ui/components-react/src/components-react/common/selection/SelectionHandler.ts#L454

And it detects changes only when `updateDragAction` is called with different item that previous or initial. So first emission will always contains the initial item for which `DragAction` was created and first different item that is handled by `DragAction`.

Reply:
Yes